README.md
[![License GPL 3][badge-license]](http://www.gnu.org/licenses/gpl-3.0.txt)
[![Build Status](https://travis-ci.org/tonini/overseer.el.svg?branch=master)](https://travis-ci.org/tonini/overseer.el)

# overseer.el

> Ert-runner Integration Into Emacs

The [ert-runner](https://github.com/rejeep/ert-runner.el) is a great way to run your ert tests. Overseer integrates this tool into emacs and let's you using it inside your lovely editor.

***

- [Installation](#installation)
  - [ELPA](#installation-via-packageel)
  - [Via el-get](#via-el-get)
  - [Manual](#manual)
- [Usage](#usage)
  - [Interactive commands](#interactive-commands)
- [Contributing](#contributing)
- [License](#license)

## Installation

### Installation via package.el

`package.el` is the built-in package manager in Emacs.

Overseer is available on the three major community maintained repositories -
[MELPA STABLE](melpa-stable.milkbox.net), [MELPA](http://melpa.milkbox.net) and [Marmalade](https://marmalade-repo.org/).

You can install `Overseer` with the following commnad:

<kbd>M-x package-install [RET] overseer [RET]</kbd>

or by adding this bit of Emacs Lisp code to your Emacs initialization file
(`.emacs` or `init.el`):

```el
(unless (package-installed-p 'overseer)
  (package-install 'overseer))
```

If the installation doesn't work try refreshing the package list:

<kbd>M-x package-refresh-contents [RET]</kbd>

Keep in mind that MELPA packages are built automatically from
the `master` branch, meaning bugs might creep in there from time to
time. Never-the-less, installing from MELPA is the recommended way of
obtaining Overseer, as the `master` branch is normally quite stable and
"stable" (tagged) builds are released somewhat infrequently.

With the most recent builds of Emacs, you can pin Overseer to always
use MELPA Stable by adding this to your Emacs initialization:

```el
(add-to-list 'package-pinned-packages '(overseer . "melpa-stable") t)
```

### Via el-get

[el-get](https://github.com/dimitri/el-get) is another popular package manager for Emacs. If you're an el-get
user just do <kbd>M-x el-get-install [RET] overseer [RET]</kbd>.

### Manual

You can install Overseer manually by placing it on your `load-path` and
`require` ing it. Many people favour the folder `~/.emacs.d/vendor`.

```el
(add-to-list 'load-path "~/.emacs.d/vendor/")
(require 'overseer)
```

## Usage

The `ert-runner` will always run in the context of the current `emacs lisp project`.

The `overseer-mode` is enabled via `emacs-lisp-mode-hook` by default for emacs lisp test files. (`*-test.el`)

### Interactive Commands

The `overseer-mode` minor mode will be automaticly enable the following keybindings:

Keybinding           | Description
---------------------|---------------
<kbd>C-c , t</kbd>   | Runs `cask exec ert-runner`. `overseer-test`
<kbd>C-c , b</kbd>   | Runs `cask exec ert-runner` with the current buffer file as argument. `overseer-test-this-buffer`
<kbd>C-c , f</kbd>   | Open a prompt to run `cask exec ert-runner` with a custom file as arguments. `overseer-test-file`
<kbd>C-c , g</kbd>   | Runs `cask exec ert-runner -t` with given tags (example: `indentation,syntax`). `overseer-test-this-buffer`
<kbd>C-c , p</kbd>   | Open a prompt to run `cask exec ert-runner` with custom arguments. `overseer-test-prompt`
<kbd>C-c , h</kbd>   | Runs `cask exec ert-runner --help`. `overseer-help`
<kbd>C-c , d</kbd>   | Runs `cask exec ert-runner --debug`. `overseer-help`
<kbd>C-c , v</kbd>   | Runs `cask exec ert-runner --verbose`. `overseer-help`
<kbd>C-c , q</kbd>   | Runs `cask exec ert-runner --quiet`. `overseer-help`

## Contributing

Contributions are very welcome!

1. Fork overseer.el
2. Create a topic branch - `git checkout -b my_branch`
4. Push to your branch - `git push origin my_branch`
5. Send me a pull-request for your topic branch
6. That's it!

## License

Copyright © 2014-2015 Samuel Tonini and
[contributors](https://github.com/tonini/overseer.el/contributors).

Distributed under the GNU General Public License, version 3

[badge-license]: https://img.shields.io/badge/license-GPL_3-green.svg?style=flat


CHANGELOG.md
# Changelog

## v0.3.0 (2015-06-08)

### Changes

  * Enable Overseer mode just in emacs lisp test files `-test.el`.

### Enhancements

  * Add keybinding for `overseer-test-file` which prompts for test file running.

## v0.2.0 (2015-01-20)

### Enhancements

  * No need to establish the root directory separate
  * Remove error link options
  * Error handling don't need be made inside the function
  * Refine codebase

### Bugfixes

  * Fix typo in function call (Nicolas Lamirault)

### Enhancements

  * Implement `overseer-test-file` function.

## v0.1.0 (2014-12-30)


test/test-helper.el
(require 'f)

(defvar overseer-test-path
  (f-parent (f-this-file)))

(defvar overseer-root-path
  (f-parent overseer-test-path))

(defvar overseer-test-path
  (f-dirname (f-this-file)))

(defvar overseer-sandbox-path
  (file-name-as-directory (f-expand "sandbox" overseer-test-path)))

(defmacro within-sandbox (&rest body)
  "Evaluate BODY in an empty temporary directory."
  `(let ((default-directory overseer-sandbox-path))
     (when (f-dir? overseer-sandbox-path)
       (f-delete overseer-sandbox-path :force))
     (f-mkdir overseer-sandbox-path)
     ,@body))

(require 'overseer (f-expand "overseer" overseer-root-path))
(require 'ert)

(provide 'test-helper)


test/overseer-test.el
;;; overseer-test.el --- Test suite for Overseer functionality.

;; Copyright © 2015 Samuel Tonini
;;
;; Author: Samuel Tonini <tonini.samuel@gmail.com>

;; This file is not part of GNU Emacs.

;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Test suite for Overseer functionality.

;;; Code:

(require 'test-helper)

(ert-deftest test-root-directory/with-root-indicator ()
  (within-sandbox
   (f-touch "Cask")
   (should (equal (overseer-project-root) (f-slash overseer-sandbox-path)))))

(ert-deftest test-if-the-current-buffer-file-is-a-test-file ()
  "Should return t if visited file is a test file"
  (within-sandbox
   (f-touch "lisp-test.el")
   (find-file "lisp-test.el")
   (should (overseer--current-buffer-test-file-p))))

(provide 'overseer-test)

;;; overseer-test.el ends here


test/ert-runner-mode-test.el
(ert-deftest test-root-directory/with-root-indicator ()
  (within-sandbox "lisp/path"
                  (f-touch "../../Cask")
                  (should (equal (ert-runner-mode-project-root) ert-runner-mode-sandbox-path))))


ert.el
;;; ert.el --- ert-runner emacs integration

;; Copyright © 2014 Samuel Tonini
;;
;; Author: Samuel Tonini <tonini.samuel@gmail.com>

;; URL: http://www.github.com/tonini/ert.el
;; Version: 0.1.0
;; Package-Requires: ((emacs "24"))
;; Keywords:

;; This file is not part of GNU Emacs.

;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;  ert-runner emacs integration

;;; Code:


(provide 'ert)

;;; ert.el ends here


ert-runner-mode.el
;;; ert-runner-mode.el --- ert-runner emacs integration

;; Copyright © 2014 Samuel Tonini
;;
;; Author: Samuel Tonini <tonini.samuel@gmail.com>

;; URL: http://www.github.com/tonini/ert.el
;; Version: 0.1.0
;; Package-Requires: ((emacs "24") (pkg-info "0.4"))
;; Keywords:

;; This file is not part of GNU Emacs.

;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;  ert-runner emacs integration

;;; Code:

(require 'compile)
(require 'ansi-color)

(defvar ert-runner-mode-command "cask exec ert-runner"
  "")

(defvar ert-runner-mode--project-root-indicators
  '("Cask")
  "list of file-/directory-names which indicate a root of a emacs lisp project")

(defun ert-runner-mode-project-root ()
  (let ((file (file-name-as-directory (expand-file-name default-directory))))
    (ert-runner-mode--project-root-identifier file ert-runner-mode--project-root-indicators)))

(defun ert-runner-mode--project-root-identifier (file indicators)
  (let ((root-dir (if indicators (locate-dominating-file file (car indicators)) nil)))
    (cond (root-dir (directory-file-name (expand-file-name root-dir)))
          (indicators (ert-runner-mode--project-root-identifier file (cdr indicators)))
          (t nil))))

(defun ert-runner-mode--establish-root-directory ()
  "Set the default-directory to the emacs lisp package project root."
  (let ((project-root (ert-runner-mode-project-root)))
    (if (not project-root)
        (error "Couldn't find any project root")
      (setq default-directory project-root))))

(defvar ert-runner-mode--buffer-name nil
  "Used to store compilation name so recompilation works as expected.")
(make-variable-buffer-local 'ert-runner-mode--buffer-name)

(defvar ert-runner-mode--error-link-options
  '(ert-runner-mode "\\([-A-Za-z0-9./_]+\\):\\([0-9]+\\)\\(: warning\\)?" 1 2 nil (3) 1)
  "File link matcher for `compilation-error-regexp-alist-alist' (matches path/to/file:line).")

(defun ert-runner-mode--kill-any-orphan-proc ()
  "Ensure any dangling buffer process is killed."
  (let ((orphan-proc (get-buffer-process (buffer-name))))
    (when orphan-proc
      (kill-process orphan-proc))))

(define-compilation-mode ert-runner-buffer-mode "ert-runner"
  "ert-runner compilation mode."
  (progn
    (font-lock-add-keywords nil
                            '(("^Finished in .*$" . font-lock-string-face)
                              ("^ert-runner.*$" . font-lock-string-face)))
    ;; Set any bound buffer name buffer-locally
    (setq ert-runner-mode--buffer-name ert-runner-mode--buffer-name)
    (set (make-local-variable 'kill-buffer-hook)
         'ert-runner-mode--kill-any-orphan-proc)))

(defvar ert-runner-mode--save-buffers-predicate
  (lambda ()
    (not (string= (substring (buffer-name) 0 1) "*"))))

(defun ert-runner-mode--handle-compilation-once ()
  (remove-hook 'compilation-filter-hook 'ert-runner-mode--handle-compilation-once t)
  (delete-matching-lines "\\(-*- mode:\\|^$\\|karma run\\|Loading config\\|--no-single-run\\|Karma finished\\|Karma started\\|karma-compilation;\\)"
                         (point-min) (point)))

(defun ert-runner-mode--handle-compilation ()
  (ansi-color-apply-on-region (point-min) (point-max)))

(defun ert-runner-mode-compilation-run (cmdlist buffer-name)
  "Run CMDLIST in `buffer-name'.
Returns the compilation buffer.
Argument BUFFER-NAME for the compilation."
  (save-some-buffers (not compilation-ask-about-save) ert-runner-mode--save-buffers-predicate)
  (let* ((ert-runner-mode--buffer-name buffer-name)
         (compilation-filter-start (point-min)))
    (with-current-buffer
        (compilation-start (mapconcat 'concat cmdlist " ")
                           'ert-runner-buffer-mode
                           (lambda (b) ert-runner-mode--buffer-name))
      (setq-local compilation-error-regexp-alist-alist
                  (cons ert-runner-mode--error-link-options compilation-error-regexp-alist-alist))
      (setq-local compilation-error-regexp-alist (cons 'ert-runner-mode compilation-error-regexp-alist))
      (add-hook 'compilation-filter-hook 'ert-runner-mode--handle-compilation nil t)
      (add-hook 'compilation-filter-hook 'ert-runner-mode--handle-compilation-once nil t))))

(defun ert-runner-mode--flatten (alist)
  (cond ((null alist) nil)
        ((atom alist) (list alist))
        (t (append (ert-runner-mode--flatten (car alist))
                   (ert-runner-mode--flatten (cdr alist))))))


(defun ert-runner-mode--build-runner-cmdlist (command)
  "Build the commands list for the runner."
  (remove "" (ert-runner-mode--flatten
              (list (if (stringp command)
                        (split-string command)
                      command)))))

(defun ert-runner-mode-run ()
  (interactive)
  (ert-runner-mode-execute (list "")
                 "*ert-runner*"))

(defun ert-runner-mode-execute (cmdlist buffer-name)
  "Run a karma command."
  (let ((old-directory default-directory))
    (ert-runner-mode--establish-root-directory)
    (cd default-directory)
    (ert-runner-mode-compilation-run (ert-runner-mode--build-runner-cmdlist (list ert-runner-mode-command cmdlist))
                           buffer-name)
    (cd old-directory)))


;;;###autoload
(defun ert-runner-mode-version (&optional show-version)
  "Get the ert-runner-mode version as string.

If called interactively or if SHOW-VERSION is non-nil, show the
version in the echo area and the messages buffer.

The returned string includes both, the version from package.el
and the library version, if both a present and different.

If the version number could not be determined, signal an error,
if called interactively, or if SHOW-VERSION is non-nil, otherwise
just return nil."
  (interactive (list t))
  (let ((version (pkg-info-version-info 'ert-runner-mode)))
    (when show-version
      (message "ert-runner-mode version: %s" version))
    version))

(provide 'ert-runner-mode)

;;; ert-runner-mode.el ends here


overseer.el
;;; overseer.el --- Ert-runner Integration Into Emacs

;; Copyright © 2014-2015 Samuel Tonini
;;
;; Author: Samuel Tonini <tonini.samuel@gmail.com>

;; URL: http://www.github.com/tonini/overseer.el
;; Version: 0.3.0
;; Package-Requires: ((emacs "24") (dash "2.10.0") (pkg-info "0.4"))
;; Keywords:

;; This file is not part of GNU Emacs.

;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Ert-runner Integration Into Emacs.

;;; Code:

(require 'compile)
(require 'dash)
(require 'pkg-info)
(require 'ansi-color)

(defgroup overseer nil
  "Ert-runner Integration Into Emacs."
  :prefix "overseer-"
  :group 'applications
  :link '(url-link :tag "Github" "https://github.com/tonini/overseer.el")
  :link '(emacs-commentary-link :tag "Commentary" "overseer"))

;; Variables

(defcustom overseer-command "cask exec ert-runner"
  "The shell command for ert-runner."
  :type 'string
  :group 'overseer)

(defvar overseer-buffer-name "*overseer*"
  "Name of the overseer buffer.")

(defvar overseer--buffer-name nil
  "Used to store compilation name so recompilation works as expected.")
(make-variable-buffer-local 'overseer--buffer-name)

(defvar overseer--project-root-indicators
  '("Cask")
  "Files which indicate a root of a emacs lisp package.")

(defvar overseer--save-buffers-predicate
  (lambda ()
    (not (string= (substring (buffer-name) 0 1) "*"))))

;; Private functions

(defun overseer--build-runner-cmdlist (command)
  "Build the arguments list for the runner with COMMAND."
  (remove "" (-flatten
              (list (if (stringp command)
                        (split-string command)
                      command)))))

(defun overseer--handle-ansi-color ()
  (ansi-color-apply-on-region (point-min) (point-max)))

(defun overseer--project-root-identifier (file indicators)
  (let ((root-dir (if indicators (locate-dominating-file file (car indicators)) nil)))
    (cond (root-dir (f-slash (directory-file-name (expand-file-name root-dir))))
          (indicators (overseer--project-root-identifier file (cdr indicators)))
          (t nil))))

(defun overseer--kill-any-orphan-proc ()
  "Ensure any dangling buffer process is killed."
  (let ((orphan-proc (get-buffer-process (buffer-name))))
    (when orphan-proc
      (kill-process orphan-proc))))

(defun overseer--test-file (filename)
  "Run ert-runner with the current FILENAME as argument."
  (overseer-execute (list (expand-file-name filename))))

(defun overseer--current-buffer-test-file-p ()
  "Return t if the current buffer is a test file."
  (string-match "-test\.el$" (or (buffer-file-name) "")))

;; Public functions

(defun overseer-project-root ()
  "Return path to the current emacs lisp package root directory."
  (let ((file (file-name-as-directory (expand-file-name default-directory))))
    (overseer--project-root-identifier file overseer--project-root-indicators)))

(define-compilation-mode overseer-buffer-mode "ert-runner"
  "Overseer compilation mode."
  (progn
    (font-lock-add-keywords nil
                            '(("^Finished in .*$" . font-lock-string-face)
                              ("^ert-runner.*$" . font-lock-string-face)))
    ;; Set any bound buffer name buffer-locally
    (setq overseer--buffer-name overseer--buffer-name)
    (set (make-local-variable 'kill-buffer-hook)
         'overseer--kill-any-orphan-proc)))

(defun overseer-compilation-run (cmdlist buffer-name)
  "Run CMDLIST in BUFFER-NAME and returns the compilation buffer."
  (save-some-buffers (not compilation-ask-about-save) overseer--save-buffers-predicate)
  (let* ((overseer--buffer-name buffer-name)
         (compilation-filter-start (point-min)))
    (with-current-buffer
        (compilation-start (mapconcat 'concat cmdlist " ")
                           'overseer-buffer-mode
                           (lambda (b) overseer--buffer-name))
      (set (make-local-variable 'compilation-error-regexp-alist) (cons 'overseer compilation-error-regexp-alist))
      (add-hook 'compilation-filter-hook 'overseer--handle-ansi-color nil t))))

(defun overseer-test ()
  "Run ert-runner."
  (interactive)
  (overseer-execute '()))

(defun overseer-help ()
  "Run ert-runner with --help as argument."
  (interactive)
  (overseer-execute '("--help")))

(defun overseer-test-this-buffer ()
  "Run ert-runner with the current `buffer-file-name' as argument."
  (interactive)
  (if (overseer--current-buffer-test-file-p)
      (overseer-execute (list (buffer-file-name)))
    (message (format "%s is no test file."
                     (file-name-nondirectory (buffer-file-name))))))

(defun overseer-test-file (filename)
  "Run `overseer--test-file' with the FILENAME."
  (interactive "Fmix test: ")
  (overseer--test-file filename))

(defun overseer-test-debug ()
  "Run ert-runner with --debug as argument."
  (interactive)
  (overseer-execute '("--debug")))

(defun overseer-test-verbose ()
  "Run ert-runner with --verbose as argument."
  (interactive)
  (overseer-execute '("--verbose")))

(defun overseer-test-quiet ()
  "Run ert-runner with --quiet as argument."
  (interactive)
  (overseer-execute '("--quiet")))

(defun overseer-test-tags (tags)
  "Run ert-runner for the given TAGS."
  (interactive "Mert-runner -t: ")
  (overseer-execute (list "-t" tags)))

(defun overseer-test-prompt (command)
  "Run ert-runner with custom arguments."
  (interactive "Mert-runner: ")
  (overseer-execute (list command)))

(defun overseer-execute (cmdlist)
  "Execute an ert-runner with CMDLIST as arguments."
  (let ((default-directory (overseer-project-root)))
    (overseer-compilation-run (overseer--build-runner-cmdlist (list overseer-command cmdlist))
                              overseer-buffer-name)))

;;;###autoload
(defun overseer-version (&optional show-version)
  "Get the Overseer version as string.

If called interactively or if SHOW-VERSION is non-nil, show the
version in the echo area and the messages buffer.

The returned string includes both, the version from package.el
and the library version, if both a present and different.

If the version number could not be determined, signal an error,
if called interactively, or if SHOW-VERSION is non-nil, otherwise
just return nil."
  (interactive (list t))
  (let ((version (pkg-info-version-info 'overseer)))
    (when show-version
      (message "Overseer version: %s" version))
    version))

(defvar overseer-mode-map
  (let ((map (make-sparse-keymap)))
    (define-key map (kbd "C-c , t") 'overseer-test)
    (define-key map (kbd "C-c , b") 'overseer-test-this-buffer)
    (define-key map (kbd "C-c , f") 'overseer-test-file)
    (define-key map (kbd "C-c , g") 'overseer-test-tags)
    (define-key map (kbd "C-c , p") 'overseer-test-prompt)
    (define-key map (kbd "C-c , d") 'overseer-test-debug)
    (define-key map (kbd "C-c , q") 'overseer-test-quiet)
    (define-key map (kbd "C-c , v") 'overseer-test-verbose)
    (define-key map (kbd "C-c , h") 'overseer-help)
    map)
  "The keymap used when `overseer-mode' is active.")

;;;###autoload
(define-minor-mode overseer-mode
  "Minor mode for emacs lisp files to test through ert-runner.

Key bindings:
\\{overseer-mode-map}"
  nil
  " overseer"
  :group 'overseer
  :global nil
  :keymap 'overseer-mode-map)

;;;###autoload
(defun overseer-enable-mode ()
  (if (overseer--current-buffer-test-file-p)
      (overseer-mode)))

;;;###autoload
(dolist (hook '(emacs-lisp-mode-hook))
  (add-hook hook 'overseer-enable-mode))

(provide 'overseer)

;;; overseer.el ends here
